ffi/src/ffi_tracing.rs

```diff
@@ -0,0 +1,555 @@
+//! FFI functions to allow engines to receive log and tracing events from kernel
+
+use std::sync::{Arc, Mutex};
+use std::{fmt, io};
+
+use delta_kernel::{DeltaResult, Error};
+use tracing::{
+    field::{Field as TracingField, Visit},
+    Event as TracingEvent, Subscriber,
+};
+use tracing_subscriber::fmt::MakeWriter;
+use tracing_subscriber::{filter::LevelFilter, layer::Context, registry::LookupSpan, Layer};
+
+use crate::{kernel_string_slice, KernelStringSlice};
+
+/// Definitions of level verbosity. Verbose Levels are "greater than" less verbose ones. So
+/// Level::ERROR is the lowest, and Level::TRACE the highest.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub enum Level {
+    ERROR = 0,
+    WARN = 1,
+    INFO = 2,
+    DEBUG = 3,
+    TRACE = 4,
+}
+
+impl Level {
+    fn is_valid(value: u32) -> bool {
+        static VALID_VALUES: &[u32] = &[0, 1, 2, 3, 4];
+        VALID_VALUES.contains(&value)
+    }
+}
+
+impl From<&tracing::Level> for Level {
+    fn from(value: &tracing::Level) -> Self {
+        match *value {
+            tracing::Level::TRACE => Level::TRACE,
+            tracing::Level::DEBUG => Level::DEBUG,
+            tracing::Level::INFO => Level::INFO,
+            tracing::Level::WARN => Level::WARN,
+            tracing::Level::ERROR => Level::ERROR,
+        }
+    }
+}
+
+impl From<Level> for LevelFilter {
+    fn from(value: Level) -> Self {
+        match value {
+            Level::TRACE => LevelFilter::TRACE,
+            Level::DEBUG => LevelFilter::DEBUG,
+            Level::INFO => LevelFilter::INFO,
+            Level::WARN => LevelFilter::WARN,
+            Level::ERROR => LevelFilter::ERROR,
+        }
+    }
+}
+
+/// An `Event` can generally be thought of a "log message". It contains all the relevant bits such
+/// that an engine can generate a log message in its format
+#[repr(C)]
+pub struct Event {
+    /// The log message associated with the event
+    message: KernelStringSlice,
+    /// Level that the event was emitted at
+    level: Level,
+    /// A string that specifies in what part of the system the event occurred
+    target: KernelStringSlice,
+    /// source file line number where the event occurred, or 0 (zero) if unknown
+    line: u32,
+    /// file where the event occurred. If unknown the slice `ptr` will be null and the len will be 0
+    file: KernelStringSlice,
+}
+
+pub type TracingEventFn = extern "C" fn(event: Event);
+
+/// Enable getting called back for tracing (logging) events in the kernel. `max_level` specifies
+/// that only events `<=` to the specified level should be reported.  More verbose Levels are "greater
+/// than" less verbose ones. So Level::ERROR is the lowest, and Level::TRACE the highest.
+///
+/// Note that setting up such a call back can only be done ONCE. Calling any of
+/// `enable_event_tracing`, `enable_log_line_tracing`, or `enable_formatted_log_line_tracing` more
+/// than once is a no-op.
```

Comment:
Should this also be part of the safety contract? 
(seems like an unusually sharp edge to have inherited from rust code)

Reply:
Ah, I didn't read very carefully... the error case becomes a no-op, so no safety issues whatsoever.

Reply:
Yeah. We should probably discuss if we should do anything other than just "return false" if you call twice, but for now it's safe. I could turn it into a proper `ExternResult` with more info, but had the feeling that was just more overhead for the engine which probably just wants to know if it set up logging okay or not.